### ISSUE TITLE
sorted(by:) works, sorted(using: KeyPathComparator) fails because it isn't sendable.

### ISSUE THREAD

**Issue by jasonbobier** [swiftlang/swift#75852]
### Description

KeyPathComparator use fails because the KeyPath isn't sendable, but both the object and the value are sendable.

### Reproduction

```swift
@MainActor
protocol Child {
	var string: String { get }
}

@MainActor
protocol Parent {
	var children: [Child] { get }
}

extension Parent {
	var sortedValues: [Child] {
		children.sorted(by: {$0.string < $1.string})	// works
		children.sorted(using: KeyPathComparator(\.string))	// fails
	}
}
```

### Expected behavior

These would both work.

### Environment

Xcode Version 16.0 beta 5 (16A5221g)

**Comment by mattneub** [comment]
This is similar to https://github.com/swiftlang/swift/issues/57560. I do not understand why that one was closed. The problem it describes has _not_ gone away with Swift 6.2, Xcode 26, and approachable concurrency with default main actor isolation. Here's an example:

```
struct Person: Sendable {
    let firstName: String
    let lastName: String
}
struct PersonStringComparator: SortComparator {
    let keyPath: KeyPath<Person, String> // WARNING
    var order: SortOrder
    init(_ keyPath: KeyPath<Person, String>, order: SortOrder = .forward) {
        self.keyPath = keyPath
        self.order = order
    }
    func compare(_ lhs: Person, _ rhs: Person) -> ComparisonResult {
        let options: String.CompareOptions = order == .forward ? [] : [.backwards]
        return lhs[keyPath: keyPath].compare(rhs[keyPath: keyPath], options: options)
    }
}
```

The warning reads: "Stored property `keyPath` of 'Sendable'-conforming struct 'PersonStringComparator' has non-Sendable type `KeyPath<Person, String>`". 

I do not know whether the warning is correct; I'm not having any actual trouble using the comparator.

**Comment by al45tair** [comment]
`KeyPath` apparently used to be unconditionally `Sendable`, but this caused a number of problems where non-`Sendable` state was inadvertently captured by a `KeyPath`.

This behaviour was changed in [SE-0418 Inferring Sendable for methods and key path literals](https://github.com/swiftlang/swift-evolution/blob/main/proposals/0418-inferring-sendable-for-methods.md), which states:

> **It is important to note that under the proposed rule all of the declarations that do not explicitly specify a Sendable requirement alongside key path type are treated as non-Sendable**

If you want a `KeyPath` to be `Sendable`, you can declare it with an `& Sendable` constraint, e.g.

```
struct PersonStringComparator: SortComparator {
    let keyPath: KeyPath<Person, String> & Sendable
    var order: SortOrder
    init(_ keyPath: KeyPath<Person, String> & Sendable, order: SortOrder = .forward) {
        self.keyPath = keyPath
        self.order = order
    }
    ...
}
```

**Comment by al45tair** [comment]
As for the OP's situation, I think the reason that fails is that `Child` is `@MainActor`, so `Child.string` is _not_ `Sendable`.

**Comment by al45tair** [comment]
This all appears to be working as expected, so closing.

**Comment by mattneub** [comment]
@al45tair The trouble with that is that you cannot then _use_ the sort comparator as a sort comparator. Example:

        var array: [Person] = [
            .init(firstName: "Harpo", lastName: "Marx"),
            .init(firstName: "Groucho", lastName: "Marx"),
            .init(firstName: "Groucho", lastName: "Aardvark")
        ]
        let comparisons: [PersonStringComparator] = [.init(\.lastName), .init(\.firstName)]
        array.sort(using: comparisons)

This results in "Type `KeyPath<Person, String>` does not conform to the 'Sendable' protocol" — and this time it's an error, not a warning, so we are dead in the water.

The entire reason that I want to revive this issue is that sort comparators remain unusable because of this problem with key paths.

**Comment by mattneub** [comment]
I request that this be reopened and left open until this conundrum is fixed at a language level.

**Comment by al45tair** [comment]
@mattneub Have you tried adding `& Sendable` to your `PersonStringComparator`, as described above?

**Comment by mattneub** [comment]
@al45tair Yes, that's what I'm responding to. Here's a full reproducible example:

```
struct Person {
    let firstName: String
    let lastName: String
}

struct PersonStringComparator: SortComparator {
    let keyPath: KeyPath<Person, String> & Sendable
    var order: SortOrder
    init(_ keyPath: KeyPath<Person, String> & Sendable, order: SortOrder = .forward) {
        self.keyPath = keyPath
        self.order = order
    }
    func compare(_ lhs: Person, _ rhs: Person) -> ComparisonResult {
        let options: String.CompareOptions = order == .forward ? [] : [.backwards]
        return lhs[keyPath: keyPath].compare(rhs[keyPath: keyPath], options: options)
    }
}
```

        // elsewhere
        var array: [Person] = [
            .init(firstName: "Harpo", lastName: "Marx"),
            .init(firstName: "Groucho", lastName: "Marx"),
            .init(firstName: "Groucho", lastName: "Aardvark")
        ]
        let comparisons: [PersonStringComparator] = [.init(\.lastName), .init(\.firstName)]
        array.sort(using: comparisons)

**Comment by al45tair** [comment]
@mattneub Could you open a *new* issue with just that example code, since I don't think that's exactly the problem the OP was reporting and I'd rather keep the two separate.
